File: packages/shared/utils/ephemeral-deadline.ts
/**
 * L'ÉCHÉANCE D'UN MESSAGE ÉPHÉMÈRE — **le point 6 du contrat du fil (#7451)**,
 * écrit UNE fois pour les clients TypeScript.
 *
 * Directive porteur 2026-09-22 : « les messages avec temps décompté ne doivent
 * décompter que lorsque l'utilisateur l'a reçu ». Jusqu'ici le client calculait
 * l'échéance à l'ENVOI (`CoreModels.swift:982-984`, recopiée telle quelle par
 * la passerelle) : un message envoyé à quelqu'un de hors ligne avait déjà brûlé
 * la moitié de sa vie quand il arrivait.
 *
 * ## LES DEUX SOURCES, ET POURQUOI ON PREND LA PLUS PROCHE
 *
 * 1. **L'échéance SERVIE** — `expiresAt` par lecteur sur REST (contrat point 3)
 *    ou l'événement `message:countdown-started` (point 5). Elle fait autorité :
 *    c'est l'heure à laquelle le serveur CESSERA de servir le message à ce
 *    lecteur.
 * 2. **La RÉCEPTION LOCALE + `ephemeralDuration`** — le repli, et le cas
 *    NOMINAL du temps réel : `message:new` ne porte plus d'`expiresAt` pour un
 *    éphémère (point 4), une diffusion en room ne pouvant pas être différente
 *    pour chaque lecteur.
 *
 * On retient la **plus proche**, jamais la plus tardive : les deux horloges
 * (celle du serveur, celle de l'appareil) dérivent l'une de l'autre, et de ces
 * deux erreurs possibles une seule est acceptable — montrer le message un
 * instant de moins que promis, jamais un instant de plus.
 *
 * ## `awaiting-reception` N'EST PAS `none`
 *
 * Un expéditeur dont personne n'a encore reçu le message n'a AUCUNE échéance :
 * la sienne se pose quand la première réception remonte (point 5, la room
 * `user:<expéditeur>`). Le dire « pas d'éphémère » (`none`) effacerait de son
 * écran la protection qu'il vient de choisir ; le dire « échu » l'effacerait
 * tout court. Il voit donc sa DURÉE, sans décompte — exactement ce que le
 * contrat appelle « en attente de réception ».
 *
 * Rien de ce fichier ne lit l'horloge : il rend une ÉCHÉANCE, jamais un
 * « reste-t-il du temps ». C'est ce qui permet à une horloge partagée (web) ou
 * à un `Text(timerInterval:)` (iOS) de rendre le décompte sans qu'aucune
 * minuterie n'appartienne à ce calcul.
 */

export type EphemeralDeadlineInput = {
  /** Secondes entières, > 0 — `Message.ephemeralDuration`, servi par REST, `message:new` et push. */
  readonly ephemeralDuration?: number | null;
  /** L'échéance SERVIE pour CE lecteur : `expiresAt` (REST) ou `message:countdown-started`. */
  readonly servedExpiresAt?: Date | string | number | null;
  /** L'instant, horloge LOCALE, où ce client a vu le message pour la première fois. */
  readonly receivedAtMs?: number | null;
  /** L'expéditeur n'a pas de « réception » : son échéance ne peut venir que du serveur. */
  readonly isMine: boolean;
};

export type EphemeralDeadline =
  | { readonly state: 'none' }
  | { readonly state: 'awaiting-reception'; readonly durationSeconds: number }
  | { readonly state: 'scheduled'; readonly expiresAtMs: number };

function durationSecondsOf(raw: number | null | undefined): number | null {
  if (raw === null || raw === undefined) return null;
  if (!Number.isFinite(raw)) return null;
  const seconds = Math.floor(raw);
  return seconds > 0 ? seconds : null;
}

function servedMsOf(raw: Date | string | number | null | undefined): number | null {
  if (raw === null || raw === undefined) return null;
  const ms = raw instanceof Date ? raw.getTime() : new Date(raw).getTime();
  return Number.isFinite(ms) ? ms : null;
}

export function ephemeralDeadline(input: EphemeralDeadlineInput): EphemeralDeadline {
  const durationSeconds = durationSecondsOf(input.ephemeralDuration);
  const served = servedMsOf(input.servedExpiresAt);

  /**
   * LA RÉCEPTION NE VAUT QUE POUR UN DESTINATAIRE. Lire l'horloge locale de
   * l'expéditeur donnerait « envoi + durée » — précisément le calcul que la
   * directive retire.
   */
  const receivedAtMs = input.receivedAtMs;
  const local =
    input.isMine || durationSeconds === null || receivedAtMs === null || receivedAtMs === undefined
      ? null
      : receivedAtMs + durationSeconds * 1000;

  if (served !== null && local !== null) {
    return { state: 'scheduled', expiresAtMs: Math.min(served, local) };
  }
  if (served !== null) return { state: 'scheduled', expiresAtMs: served };
  if (local !== null) return { state: 'scheduled', expiresAtMs: local };
  if (durationSeconds !== null) return { state: 'awaiting-reception', durationSeconds };
  return { state: 'none' };
}

/**
 * LA DERNIÈRE MINUTE — **le compteur n'apparaît que là** (#7468 web, #7467 iOS).
 *
 * Précision du porteur, 2026-09-22 : « l'éphémère est la flamme avec la
 * configuration de durée par défaut ! Ce qu'il faudrait c'est d'afficher le
 * compteur de l'éphémère dans la conversation uniquement quand on est déjà à
 * 1 min et moins de sa destruction. »
 *
 * La valeur vit ICI, avec la règle d'échéance, parce qu'elle en est la SUITE :
 * l'échéance dit QUAND, celle-ci dit à partir de quand on l'ÉCRIT. Les deux
 * clients la partagent — un seuil recopié dans deux peaux divergerait au
 * premier ajustement.
 */
export const EPHEMERAL_COUNTER_WINDOW_SECONDS = 60;

/**
 * `true` dès que le compteur doit se peindre — et il gouverne DEUX choses que
 * rien d'autre ne relie : ce que l'œil voit (des chiffres, ou la seule flamme)
 * et ce qui s'abonne à l'horloge. Au-delà de la fenêtre, AUCUNE horloge ne bat
 * pour ce message : il n'y a rien à rafraîchir, et vingt rangées éphémères
 * réveillaient vingt fois par seconde un écran qui n'en montrait rien.
 *
 * Un reste NÉGATIF rend `true` : une échéance passée n'est pas « au-delà de la
 * minute » — c'est un message en train de disparaître, et lui remettre la
 * flamme nue serait un pas en arrière au pire moment.
 */
export function ephemeralCounterVisible(remainingMs: number): boolean {
  return remainingMs <= EPHEMERAL_COUNTER_WINDOW_SECONDS * 1000;
}

File: packages/MeeshySDK/Sources/MeeshySDK/Models/EphemeralDeadline.swift
import Foundation

/// L'échéance d'un message éphémère — **une seule règle, pour toutes les
/// surfaces** (#7452, contrat du fil #7451 point 6).
///
/// ## Deux horloges, et pourquoi la plus proche gagne
///
/// Jusqu'au 2026-09-22 l'échéance d'un éphémère naissait à l'ENVOI, chez
/// l'expéditeur (`EphemeralDuration.expiresAt` = `Date() + rawValue`), puis le
/// gateway la recopiait telle quelle pour tout le monde. Un destinataire qui
/// ouvrait la conversation quatre minutes après l'envoi d'un message de cinq
/// minutes ne le voyait vivre qu'une minute — et un destinataire hors ligne
/// pendant six minutes ne le voyait jamais. La directive porteur tranche :
/// « les messages avec temps décompté ne doivent décompter que lorsque
/// l'utilisateur l'a reçu ».
///
/// Deux sources peuvent donc dire quand un message meurt, et elles ne se
/// remplacent pas :
///
/// | source | qui la produit | quand elle manque |
/// |---|---|---|
/// | `servedExpiresAt` | le gateway, résolu POUR CE LECTEUR (`D(u)`), servi par REST et par `message:countdown-started` | tant que le serveur n'a pas enregistré la réception, ou tant que #7451 n'est pas fusionnée |
/// | réception locale `+ ephemeralDuration` | le client, à la première fois où il VOIT le message | pour l'expéditeur, tant que personne n'a reçu |
///
/// **La plus PROCHE des deux gagne.** C'est une règle de sécurité, pas
/// d'affichage : aucune des deux sources ne peut PROLONGER la vie d'un message
/// protégé. Si le serveur dit 14:32 et que l'arithmétique locale dit 14:35,
/// c'est 14:32 — et réciproquement. `min` est ici le seul opérateur qui ne
/// puisse pas trahir la protection.
///
/// `message:new` ne porte plus d'`expiresAt` pour un éphémère (contrat point 4,
/// une diffusion en room ne peut pas être différente par lecteur) : sur ce
/// chemin, seule la seconde ligne du tableau existe. C'est ce qui rend ce lot
/// indépendant de la fusion de #7451.
public enum EphemeralDeadline {

    /// **En deçà de combien de temps le COMPTEUR s'affiche** (#7467).
    ///
    /// > « afficher le compteur de l'éphémère dans la conversation uniquement
    /// > quand on est déjà à 1 min et moins de sa destruction »
    ///
    /// Avant ce seuil, la flamme SEULE dit déjà tout ce qu'il y a à savoir :
    /// ce message va disparaître. Le CHIFFRE, lui, ne devient une information
    /// qu'au moment où il devient une urgence — un message de vingt-quatre
    /// heures affichait « 23:59:58 » sous chaque bulle, une horloge qui ne dit
    /// rien pendant vingt-trois heures et fait battre une seconde pour rien.
    public static let countdownThreshold: TimeInterval = 60

    /// Ce qu'il y a à AFFICHER pour un message, à un instant donné.
    ///
    /// `notEphemeral` plutôt que `none` : `none` entrerait en collision de
    /// lecture avec `Optional.none` sur chaque site de comparaison, et la
    /// distinction « pas éphémère » / « éphémère sans échéance » est justement
    /// celle que ce type existe pour porter.
    public enum State: Equatable, Sendable {
        /// Le message n'est pas éphémère — aucun chrome, aucun décompte.
        case notEphemeral
        /// Éphémère, mais personne n'a encore reçu : la DURÉE s'affiche, sans
        /// décompte. C'est ce que voit l'expéditeur avant le premier accusé.
        case awaitingReception(duration: TimeInterval)
        /// Éphémère, l'horloge tourne jusqu'à `deadline` — mais on en est
        /// ENCORE LOIN : la flamme seule, sans chiffre (#7467).
        case running(deadline: Date)
        /// Éphémère dans sa DERNIÈRE MINUTE : la flamme ET le compteur qui
        /// défile, 0:59 → 0:00.
        ///
        /// Deux cas plutôt qu'un booléen à côté de `running` : le rendu n'est
        /// pas le même, et c'est le MODÈLE qui doit le dire. Une vue qui
        /// lirait `Date()` dans son corps pour trancher deviendrait non
        /// déterministe — donc intestable, et son `Equatable` mentirait sur ce
        /// qu'elle affiche.
        case imminent(deadline: Date)
        /// L'échéance est passée : le message quitte l'écran.
        case expired

        /// L'échéance, quand il y en a une. Évite de déballer le cas sur les
        /// sites qui n'ont besoin que de la date (ordonnancement du balayage).
        public var deadline: Date? {
            switch self {
            case .running(let deadline), .imminent(let deadline): return deadline
            case .notEphemeral, .awaitingReception, .expired: return nil
            }
        }

        /// Le compteur chiffré doit-il s'afficher ?
        ///
        /// La question se pose à la VUE, la réponse vient du modèle. C'est
        /// aussi ce qui garantit qu'aucune horloge ne bat avant la dernière
        /// minute : `Text(timerInterval:)` n'est monté que dans ce cas.
        public var showsCountdown: Bool {
            if case .imminent = self { return true }
            return false
        }

        /// Un message éphémère (quel que soit l'état de son horloge).
        public var isEphemeral: Bool { self != .notEphemeral }
    }

    /// Résout l'état d'affichage depuis les deux sources du contrat.
    ///
    /// - Parameters:
    ///   - servedExpiresAt: `D(lecteur)` servi par le gateway, ou `nil`.
    ///   - ephemeralDuration: la durée en SECONDES portée par le message.
    ///     Une valeur nulle ou négative ne fabrique aucune échéance : elle
    ///     signifierait « déjà mort à la réception », ce qu'aucun producteur
    ///     n'a le droit de dire par omission.
    ///   - localReceivedAt: la PREMIÈRE fois que cet appareil a vu le message.
    ///   - now: l'instant de référence (injecté pour les témoins).
    public static func resolve(
        servedExpiresAt: Date?,
        ephemeralDuration: Int?,
        localReceivedAt: Date?,
        now: Date = Date()
    ) -> State {
        let duration: TimeInterval? = {
            guard let ephemeralDuration, ephemeralDuration > 0 else { return nil }
            return TimeInterval(ephemeralDuration)
        }()

        let localDeadline: Date? = {
            guard let duration, let localReceivedAt else { return nil }
            return localReceivedAt.addingTimeInterval(duration)
        }()

        let candidates = [servedExpiresAt, localDeadline].compactMap { $0 }

        guard let deadline = candidates.min() else {
            guard let duration else { return .notEphemeral }
            return .awaitingReception(duration: duration)
        }

        if deadline <= now { return .expired }
        return deadline.timeIntervalSince(now) <= countdownThreshold
            ? .imminent(deadline: deadline)
            : .running(deadline: deadline)
    }
}
